src/v/datalake/translation/state_machine.cc

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/translation/state_machine.h"
+
+#include "datalake/logger.h"
+#include "datalake/translation/types.h"
+
+namespace {
+raft::replicate_options make_replicate_options() {
+    auto opts = raft::replicate_options(raft::consistency_level::quorum_ack);
+    opts.set_force_flush();
+    return opts;
+}
+
+model::record_batch_reader make_translation_state_batch(kafka::offset offset) {
+    auto val = datalake::translation::translation_state{
+      .highest_translated_offset = offset};
+    storage::record_batch_builder builder(
+      model::record_batch_type::datalake_translation_state, model::offset(0));
+    builder.add_raw_kv(std::nullopt, serde::to_iobuf(val));
```

Comment:
i think setting a key to null may impact compacted topics, but i think we should not compact batches that contribute to offset translation.